url2698a2c731_cassandra_project_cassandra_stand_alone_tgz-pJ8-eu_cassandra_sim_MySimpleDynamicScenarioSimulationJ8/src/eu/cassandra/sim/math/response/Response.java

```diff
@@ -0,0 +1,669 @@
+		// double sum = 0;
+
+		int start, start2, end, end2, duration;
+		double previousPrice, newPrice;
+
+		if (pricing.getNumberOfPenalties() > 0) {
+
+			Map<Integer, Double> percentageMap = new TreeMap<Integer, Double>();
+			ArrayList<Integer> tempList = new ArrayList<Integer>(pricing.getBases());
+			tempList.addAll(pricing.getRewards());
+
+			for (Integer index: tempList)
+				sum += pricing.getPricings(index).getGainRatio();
+
+			for (Integer index: tempList)
+				percentageMap.put(index, pricing.getPricings(index).getGainRatio()
+						/ sum);
+
+			// System.out.println("Percentage Map: " + percentageMap.toString());
+
+			for (Integer index: pricing.getPenalties()) {
+				overDiff = 0;
+				sum = 0;
+
+				tempPricing = pricing.getPricings(index);
+				start = tempPricing.getStartMinute();
+				end = tempPricing.getEndMinute();
+				previousPrice = tempPricing.getPreviousPrice();
+				newPrice = tempPricing.getCurrentPrice();
+
+				for (int i = start; i <= end; i++) {
+
+					temp = ((previousPrice * values[i]) / newPrice);
+					// System.out.println("Temp = " + temp);
+					overDiffTemp = (values[i] - temp) * awareness * sensitivity;
+					overDiff += overDiffTemp;
+					values[i] -= overDiffTemp;
+				}
+
+				// System.out.println("OverDiff for index " + index + ": " + overDiff);
+
+				for (Integer index2: tempList) {
+					start2 = pricing.getPricings(index2).getStartMinute();
+					end2 = pricing.getPricings(index2).getEndMinute();
+					duration = end2 - start2;
+					additive = overDiff * percentageMap.get(index2) / duration;
+					// System.out.println("Additive for index " + index2 + ": " +
+					// additive);
+					for (int i = start2; i < end2; i++)
+						values[i] += additive;
+				}
+
+				for (int i = 0; i < values.length; i++)
+					sum += values[i];
+
+				// System.out.println("Summary: " + sum);
+
+			}
+
+		}
+		else if (pricing.getNumberOfRewards() > 0) {
+
+			Pricing tempPricing2 = null;
+			ArrayList<Pricing> tempList = new ArrayList<Pricing>();
+
+			for (Integer index: pricing.getRewards())
+				tempList.add(pricing.getPricings(index));
+
+			Collections.sort(tempList, comp);
+
+			// System.out.println("Rewards List: " + tempList.toString());
+
+			for (int i = 0; i < tempList.size(); i++) {
+
+				tempPricing2 = tempList.get(i);
+				newPrice = tempPricing2.getCurrentPrice();
+				start2 = tempPricing2.getStartMinute();
+				end2 = tempPricing2.getEndMinute();
+				duration = end2 - start2;
+
+				for (Integer index: pricing.getBases()) {
+					overDiff = 0;
+					sum = 0;
+
+					tempPricing = pricing.getPricings(index);
+					start = tempPricing.getStartMinute();
+					end = tempPricing.getEndMinute();
+					previousPrice = tempPricing.getCurrentPrice();
+
+					for (int j = start; j <= end; j++) {
+
+						temp = newPrice * values[j] / previousPrice;
+						overDiffTemp = (values[j] - temp) * awareness * sensitivity;
+						overDiff += overDiffTemp;
+						values[j] -= overDiffTemp;
+
+					}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 436:443, 503:510.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.